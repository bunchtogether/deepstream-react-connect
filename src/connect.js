// @flow 

import DeepstreamClient from 'deepstream.io-client-js';
import type { ComponentType } from 'react';
import React from 'react';
import { subscribe } from './lib/record';

type Props = {
  [string]:[any, Array<string>]
};

type State = {
  [string]:any
};

export default (client:DeepstreamClient, parameters:{[string]:string}, Component:ComponentType<*>):ComponentType<*> => class ConnectedComponent extends React.PureComponent<Props, State> {
  static displayName = `Connected${Component.displayName || Component.name}`;

  constructor(props: Props) {
    super(props);
    this.state = {};
    this.subscriptions = {};
  }

  componentWillMount() {
    this.setSubscriptions(this.props);
  }

  componentWillReceiveProps(nextProps:Props) {
    this.unsetSubscriptions();
    this.setSubscriptions(nextProps);
  }

  componentWillUnmount() {
    this.unsetSubscriptions();
  }

  setSubscriptions(props:Props) {
    Object.keys(props).forEach((name) => {
      if (name === 'children') {
        return;
      }
      const subscription = subscribe(client, props[name][0], props[name][1], parameters);
      subscription.addCallback((value:any) => {
        if (this.state[name] !== value) {
          this.setState({
            [name]: value,
          });
        }
      });
      this.subscriptions[name] = subscription;
    });
  }

  unsetSubscriptions() {
    Object.keys(this.subscriptions).forEach((name) => this.subscriptions[name].close());
  }

  subscriptions: {
    [string]: {
      addCallback:Function,
      close: Function
    }
  };

  render() {
    return <Component {...this.state} />;
  }
};

