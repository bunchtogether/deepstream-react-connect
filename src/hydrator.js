// @flow 

import type { Element, ComponentType } from 'react';

import uuid from 'uuid';
import DeepstreamClient from 'deepstream.io-client-js';
import React, { createElement } from 'react';

export class Text extends React.Component<{value:string}> { // eslint-disable-line react/prefer-stateless-function
  render() {
    return null;
  }
}

function getElementType(element:Element<*>) {
  if (typeof (element.type) === 'string') {
    return element.type;
  }
  if (typeof (element.type.name) === 'string') {
    return element.type.name;
  }
  if (typeof (element.type.displayName) === 'string') {
    return element.type.displayName;
  }
  throw new Error('Could not get display name of element');
}

export default class Hydrator {
  client: DeepstreamClient;
  components:{[string]:ComponentType<*>};
  constructor(client:DeepstreamClient, components:Array<ComponentType<*>>) {
    this.client = client;
    this.components = {};
    components.forEach((Component) => {
      this.components[Component.name] = Component;
    });
  }
  async dehydrate(element:Element<*>):Promise<string> {
    const response = {};
    response.type = getElementType(element);
    response.props = Object.assign({}, element.props);
    response.props.key = response.props.key || uuid.v4();
    const childPromises = [];
    React.Children.map(element.props.children, (child) => {
      if (React.isValidElement(child)) {
        childPromises.push(this.dehydrate(child));
      } else {
        childPromises.push(this.dehydrate(<Text value={child} />));
      }
    });
    response.props.children = await Promise.all(childPromises);
    await new Promise((resolve, reject) => {
      this.client.record.setData(response.props.key, response, (error) => {
        if (error) {
          reject(error);
        } else {
          resolve();
        }
      });
    });
    return response.props.key;
  }
  async hydrate(key:string):Promise<Element<*> | string> {
    const response = await new Promise((resolve, reject) => {
      this.client.record.snapshot(key, (error, data) => {
        if (error) {
          reject(error);
        } else {
          resolve(data);
        }
      });
    });
    const componentType = this.components[response.type] || response.type;
    if (componentType === 'Text') {
      return response.props.value;
    }
    const children = await Promise.all(response.props.children.map(this.hydrate.bind(this)));
    delete response.props.children;
    return createElement(componentType, response.props, children);
  }
}
