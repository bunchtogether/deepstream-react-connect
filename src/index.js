// @flow 

// import React from 'react';

export const connect = <T: *>(Component:ReactClass<T>): ReactClass<T> => Component;

