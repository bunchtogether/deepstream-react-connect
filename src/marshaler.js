// @flow 

import type { Element, ComponentType } from 'react';
import React, { createElement } from 'react';
import murmurhash from 'murmurhash3js';

export type Instance = {
  type: string,
  key: string,
  props: {[string]: any},
  children: Array<Instance>
}

function getElementType(element:Element<*>) {
  if (typeof (element.type.displayName) === 'string') {
    return element.type.displayName;
  }
  if (typeof (element.type) === 'string') {
    return element.type;
  }
  if (typeof (element.type.name) === 'string') {
    return element.type.name;
  }
  throw new Error('Could not get display name of element');
}


class Text extends React.Component<{value:string}> { // eslint-disable-line react/prefer-stateless-function
  render() {
    return null;
  }
}

export default class Marshaler {
  components:{[string]:ComponentType<*>};

  constructor(components:Array<ComponentType<*>>) {
    this.components = {};
    components.forEach((Component) => {
      this.components[Component.displayName || Component.name] = Component;
    });
  }

  marshal = (element:Element<any>):Instance => {
    const children = [];
    React.Children.map(element.props.children, (child) => {
      if (React.isValidElement(child)) {
        children.push(this.marshal(child));
      } else {
        children.push(this.marshal(<Text value={child} />));
      }
    });
    const props = Object.assign({}, element.props);
    delete props.children;
    const instance:Object = {
      type: getElementType(element),
      props,
      children,
    };
    instance.key = element.key === null ? murmurhash.x64.hash128(JSON.stringify(instance)) : element.key.toString();
    return instance;
  }

  unmarshal = (instance:Instance):Element<any> => {
    const componentType = this.components[instance.type] || instance.type;
    const props = Object.assign({ key: instance.key }, instance.props);
    const children = instance.children.map((childInstance: Instance) => {
      if (!childInstance) {
        return null;
      }
      if (childInstance.type === 'Text') {
        return childInstance.props.value;
      }
      return this.unmarshal(childInstance);
    });
    return createElement(componentType, props, children);
  }
}
