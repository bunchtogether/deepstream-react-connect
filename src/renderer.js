// @flow 

import React, { createElement } from 'react';

type Instance = {
  type: string,
  props: Object,
  children: Array<Instance>
};

const KNOWN_TYPES = {

};

export const render = (instance: Instance):React.Element<*> => {
  if(KNOWN_TYPES[instance.type]) {
    return createElement(
      KNOWN_TYPES[instance.type],
      instance.props,
      instance.children.map((child) => render(child))
    )
  }
  return createElement(
    KNOWN_TYPES[instance.type],
    instance.props,
    instance.children.map((child) => render(child))
  )
};
