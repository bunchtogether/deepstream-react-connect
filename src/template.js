/*

{
  type: 'example',
  props: [[name, path], ...],
  children: listName
}

*/

