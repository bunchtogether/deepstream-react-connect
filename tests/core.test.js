// @flow 

import React from 'react';
import renderer from 'react-test-renderer';
// import ReactTestUtils from 'react-dom/test-utils';
import TestComponent from './lib/test-component';
import { connect } from '../src';

test('Wrap a component', () => {
  const ConnectedTestComponent = connect(TestComponent);
  const component = renderer.create(<ConnectedTestComponent />);
  const tree = component.toJSON();
  console.log(tree);
});
