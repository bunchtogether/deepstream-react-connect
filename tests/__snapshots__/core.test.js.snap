// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a div 1`] = `<div />`;

exports[`Wrap a component 1`] = `<div />`;
